File: 0023. Merge k Sorted Lists/main0023.java
public class main0023 {
    
}

class Solution0023 {
    public ListNode mergeKLists(ListNode[] lists) {
        ListNode rst = new ListNode();
        ListNode cur = rst;
        int len = lists.length;
        if (len == 0) {
            return null;
        } else if (len == 1) {
            return lists[0];
        }
        // PriorityQueue<ListNode> minHeap = new PriorityQueue<ListNode>(len, new Comparator<ListNode>(){
        //     @Override
        //     public int compare(ListNode L1, ListNode L2) {
        //         if (L1.val == L2.val) {
        //             return 0;
        //         }
        //         return L1.val < L2.val ? -1 : 1;
        //     }
        // });
        PriorityQueue<ListNode> minHeap = new PriorityQueue<ListNode>((l1, l2)-> l1.val - l2.val);
        // initialization
        for (int i = 0; i < len; i++) {
            if (lists[i] != null) {
                minHeap.offer(lists[i]);     
            }
        }
        // pop
        while (!minHeap.isEmpty()) {
            ListNode tmp = minHeap.poll();
            if (tmp.next != null) {
                minHeap.offer(tmp.next);
            }
            cur.next = tmp;
            cur = cur.next;
            cur.next = null;
        }
        
        // return
        return rst.next;
    }
}
// row, col
// tc: row*col*log(row)
// 3:1,1  (3-1)/2, 3/2 nums1[nums1.length / 2] + 
// 4:1,2->(4-1)/2, 4/2

File: 0023. Merge k Sorted Lists/main0023.py

from Queue import PriorityQueue
class ListNode:
    def __init__(self, val = 0, next = None):
        self.val = val
        self.next = next
    
class Solution0023:
    def mergeKLists(self, lists):
        dum = cur = ListNode(0)
        q = PriorityQueue()
        #init
        for l in lists:
            if l:
                q.put((l.val, l))
        while not q.empty():
            val, node = q.get()
            cur.next = ListNode(val)
            node = node.next
            cur = cur.next
            if node:
                q.put((node.val, node))
        return dum.next
